repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -106,20 +100,34 @@ class Session(interpreter: Interpreter)
     _statements.clear()
   }
 
+  private def changeState(newState: SessionState): Unit = {
+    synchronized {
+      _state = newState
+    }
+    stateChangedCallback(newState)
+  }
+
   private def executeCode(executionCount: Int, code: String): String = synchronized {
-    _state = SessionState.Busy()
+    changeState(SessionState.Busy())
+
+    def transitToIdle() = {
+      val executingLastStatement = executionCount == newStatementId.intValue() - 1
+      if (_statements.isEmpty || executingLastStatement) {
```

Comment:
Also how to deal with the "busy" state for user is a question. What user get will always be later than the state updated in server side, so user may get a chance to get "idle" state and do some actions, but the state is changed to "busy" in server side, and vice versa. So in such case state showing to user and related actions by user cannot really prove the actual state and action for user.